repository/src/main/java/org/apache/atlas/discovery/EntityLineageService.java

```diff
@@ -454,6 +453,15 @@ private boolean incrementAndCheckIfRelationsLimitReached(AtlasEdge atlasEdge, bo
         String                     outGuid                   = AtlasGraphUtilsV2.getIdFromVertex(outVertex);
         LineageOnDemandConstraints outGuidLineageConstraints = getAndValidateLineageConstraintsByGuid(outGuid, atlasLineageOnDemandContext);
 
+        // Special checks to handle self cyclic lineage on base entity
+        boolean areBothDirectionsRequested = baseEntityHasSelfCyclicLineage && AtlasLineageOnDemandInfo.LineageDirection.BOTH.equals(outGuidLineageConstraints.getDirection());
+        boolean skipOutputIncrement = baseEntityHasSelfCyclicLineage && isInput && areBothDirectionsRequested;
```

Comment:
it is better to check for direction.equals(INPUT) instead of isInput because even when we process on the OUTPUT side, isInput will be true while processing for the first edge of relation. this situation is not fatal.